README.md
# Graphite module for Icinga Web 2

## General Information

Enable the graphite carbon cache writer: http://docs.icinga.org/icinga2/latest/doc/module/icinga2/chapter/monitoring-basics#graphite-carbon-cache-writer

All perfdata metrics will be automatically included as graphs however if you just want a subset, the host or service then needs to have custom vars of the form vars.graphite_keys =["key1","key2"] where key1 key2 represent perfdata stats you want to see.

You can configure the graphite metric keys formats by using standard-ish icinga2 macros.
## Installation

Just extract this to your Icinga Web 2 module folder in a folder called graphite.

(Configuration -> Modules -> graphite -> enable). Check the modules config tab right there.

NB: It is best practice to install 3rd party modules into a distinct module
folder like /usr/share/icingaweb2/modules. In case you don't know where this
might be please check the module path in your Icinga Web 2 configuration.

##Configuration
There are various configuration settings to tweak how the module behaves and ensure that it aligns with how the graphite carbon cache writer is set up:

``base_url``
A fully formed url 
* *http://graphite.com/render?*

``legacy_mode``
To support older versions of the writer pre-icinga2 2.4 where true - is replaced with _ 
* *false*

``service_name_template``
Macro template for the service name 
* *icinga2.$host.name$.services.$service.name$.$service.check_command$.perfdata.$metric$.value*

``host_name_template``
Macro template for the host name 
* *icinga2.$host.name$.host.$host.check_command$.perfdata.$metric$.value*

``graphite_args_template ``
Macro template for the small image where $target$ is replaced with the metric name 
* *&target=$target$&source=0&width=300&height=120&hideAxes=true&lineWidth=2&hideLegend=true&colorList=049BAF*

``graphite_large_args_template ``
Macro template for the large image 
* *&target=$target$&source=0&width=800&height=700&colorList=049BAF&lineMode=connected*

## Customizing
As mentioned in General Information above, you can use vars.graphite_keys to limit [or define] the graphs you want to see.

### Limiting graphs
In our example there are 6 metrics available in perfdata from the check_tomcat script:
* used
* free
* max
* currentThreadCount
* maxThreads
* currentThreadsBusy

To just see the basic graphs of used heap and currentThreadCount, add the following to your Service definition.

    vars.graphite_keys = ["used", "currentThreadCount"]

### Composite Graphs
In some cases, composite graphs are more useful.  Used Heap doesn't make sense unless you know the Max Heap size.  To create some useful composite graphs, add the following to your Service definition.

    vars.graphite_keys = ["{used,max}", "{currentThreadsBusy,currentThreadCount}"]

### Graph Labels
Composite graph definitions can get long and cumbersome, so you can define Labels for each graph.  To make short, user-friendly labels for the composite graphs, add the following to your Service definition.

    vars.graphite_keys = ["{used,max}", "{currentThreadsBusy,currentThreadCount}"]
    vars.graphite_labels = ["Heap", "Threads"]

## Hats off to

This module borrows a lot from https://github.com/Icinga/icingaweb2-module-pnp4nagios

## What it looks like		

![screen shot of graphite graph](https://raw.githubusercontent.com/philiphoy/icingaweb2-module-graphite/master/Capture.PNG)


CHANGELOG.md
0.0.0.5
=========
* Added labels for metrics
* Added graphite area overload option

0.0.0.4
=========
* Removed metric_prefix from config
* Added metric prefix to macro templates
* Added metric to macro templates

0.0.0.3
=========
* Removed .value from Grapher.php and added to the macro templates, the default templates respect this change.

0.0.0.2
=========
* Changed defaults for service_name_template and host_name_template to align with the example configuration.


library/Graphite/Macro.php
<?php
/* Icinga Web 2 | (c) 2013-2015 Icinga Development Team | GPLv2+ */

namespace Icinga\Module\Graphite;

/**
 * Expand macros in string in the context of MonitoredObjects
 */
class Macro
{
    /**
     * Known icinga macros
     *
     * @var array
     */
    private static $icingaMacros = array(
        'HOSTNAME'     => 'host_name',
        'HOSTADDRESS'  => 'host_address',
        'SERVICEDESC'  => 'service_description',
        'service.name' => 'service_description'
    );

    /**
     * Return the given string with macros being resolved
     *
     * @param   string                      $input      The string in which to look for macros
     * @param   MonitoredObject|stdClass    $object     The host or service used to resolve macros
     *
     * @return  string                                  The substituted or unchanged string
     */
    public static function resolveMacros($input, $object, $legacyMode, $escape = false , $isMacro = false)
    {
        $matches = array();
        if (preg_match_all('@\$([^\$\s]+)\$@', $input, $matches)) {
            foreach ($matches[1] as $key => $value) {
                $newValue = self::resolveMacro($value, $object);
                if ($newValue !== $value) {
                    if ($escape){
                      $newValue = self::escapeMetric($newValue, $legacyMode, $isMacro);
                    }
                    $input = str_replace($matches[0][$key], $newValue, $input);
                }
            }
        }

        return $input;
    }


    /**
     * Resolve a macro based on the given object
     *
     * @param   string                      $macro      The macro to resolve
     * @param   MonitoredObject|stdClass    $object     The object used to resolve the macro
     *
     * @return  string                                  The new value or the macro if it cannot be resolved
     */
    public static function resolveMacro($macro, $object)
    {
        if (is_array($object) && array_key_exists($macro, $object)) {
            return $object[$macro];
        }

        if (array_key_exists($macro, self::$icingaMacros) && $object->{self::$icingaMacros[$macro]} !== false) {
            return $object->{self::$icingaMacros[$macro]};
        }

        if (array_key_exists($macro, $object->customvars)) {
            return $object->customvars[$macro];
        }

        $translated = self::translateTerm($macro);

        if ($translated !== NULL) {
            if (array_key_exists($translated, $object->customvars)) {
                return $object->customvars[$translated];
            }

            if  ($object->{$translated} !== false) {
               return $object->{$translated};
            }
        }

        return $macro;
    }

    public static function escapeMetric($str, $legacyMode, $ismetric)
    {
        if ($legacyMode) {
            $str=str_replace('-','_',$str);
            $str=str_replace('.','_',$str);
        } elseif (!$ismetric){
            $str=str_replace('.','_',$str);
        }

        $str=str_replace(' ','_',$str);
        $str=str_replace('\\','_',$str);
        $str=str_replace('/','_',$str);
        $str=str_replace('::','.',$str);
        return $str;
    }

    private static function translateTerm($term){

        if (substr($term, 0, strlen('host.vars.')) === 'host.vars.'){
            return str_replace('host.vars.','',$term);
        }

        if (substr($term, 0, strlen('service.vars.')) === 'service.vars.'){
            return str_replace('service.vars.','',$term);
        }

        if (substr($term, 0, strlen('service.')) === 'service.'){
            return str_replace('service.','service_',$term);
        }

        if (substr($term, 0, strlen('host.')) === 'host.'){
            return str_replace('host.','host_',$term);
        }
    }
}


library/Graphite/Grapher.php
<?php

namespace Icinga\Module\Graphite;

use Icinga\Application\Config;
use Icinga\Exception\ConfigurationError;
use Icinga\Module\Monitoring\Object\MonitoredObject;
use Icinga\Module\Monitoring\Object\Host;
use Icinga\Module\Monitoring\Object\Service;
use Icinga\Module\Monitoring\Plugin\PerfdataSet;
use Icinga\Web\Hook\GrapherHook;
use Icinga\Web\Url;

class Grapher extends GrapherHook
{
    protected $hasPreviews = true;
    protected $hasTinyPreviews = true;
    protected $graphiteConfig;
    protected $baseUrl = 'http://graphite.com/render/?';
    protected $serviceMacro = 'icinga2.$host.name$.services.$service.name$.$service.check_command$.perfdata.$metric$.value';
    protected $hostMacro = 'icinga2.$host.name$.host.$host.check_command$.perfdata.$metric$.value';
    protected $imageUrlMacro = '&target=$target$&source=0&width=300&height=120&hideAxes=true&lineWidth=2&hideLegend=true&colorList=049BAF';
    protected $largeImageUrlMacro = '&target=$target$&source=0&width=800&height=700&colorList=049BAF&lineMode=connected';
    protected $legacyMode = false;
    protected $graphiteKeys = array();
    protected $graphiteLabels = array();
    protected $areaMode = "all";
    protected $iframeWidth = "800px";
    protected $iframeHeight = "700px";

    protected function init()
    {
        $cfg = Config::module('graphite')->getSection('graphite');
        $this->baseUrl = rtrim($cfg->get('base_url', $this->baseUrl), '/');
        $this->legacyMode = filter_var($cfg->get('legacy_mode', $this->legacyMode), FILTER_VALIDATE_BOOLEAN);
        $this->serviceMacro = $cfg->get('service_name_template', $this->serviceMacro);
        $this->hostMacro = $cfg->get('host_name_template', $this->hostMacro);
        $this->imageUrlMacro = $cfg->get('graphite_args_template', $this->imageUrlMacro);
        $this->largeImageUrlMacro = $cfg->get('graphite_large_args_template', $this->largeImageUrlMacro);
        $this->iframeWidth = $cfg->get('graphite_iframe_w', $this->iframeWidth);
        $this->iframeHeight = $cfg->get('graphite_iframe_h', $this->iframeHeight);
    }

    private function parseGrapherConfig($graphite_vars)
    {
        if (!empty($graphite_vars)) {
            if (!empty($graphite_vars->area_mode)) {
                $this->areaMode = $graphite_vars->area_mode;
            }
        }
    }

    private function getKeysAndLabels($vars)
    {
        if (array_key_exists("graphite_keys", $vars)) {
            $this->graphiteKeys = $vars["graphite_keys"];
            $this->graphiteLabels = $vars["graphite_keys"];
            if (array_key_exists("graphite_labels", $vars)) {
                if (count($vars["graphite_keys"]) == count($vars["graphite_labels"])) {
                    $this->graphiteLabels = $vars["graphite_labels"];
                }
            }
        }
    }

    private function getPerfdataKeys($object)
    {
        foreach (PerfdataSet::fromString($object->perfdata)->asArray() as $pd) {
            $this->graphiteKeys[] = $pd->getLabel();
            $this->graphiteLabels[] = $pd->getLabel();
        }
    }

    private function getPreviewImage($host, $service, $metric)
    {
        if ($host != null){
            $target = Macro::resolveMacros($this->hostMacro, $host, $this->legacyMode, true);
        } elseif  ($service != null ){
            $target = Macro::resolveMacros($this->serviceMacro, $service, $this->legacyMode, true);
        } else {
           $target = '';
        }

        $target = Macro::resolveMacros($target, array("metric"=>$metric), $this->legacyMode, true, true);
        $imgUrl = $this->baseUrl . Macro::resolveMacros($this->imageUrlMacro, array("target" => $target, "areaMode" => $this->areaMode), $this->legacyMode);
        $largeImgUrl = $this->baseUrl . Macro::resolveMacros($this->largeImageUrlMacro, array("target" => $target, "areaMode" => $this->areaMode), $this->legacyMode);

        $url = Url::fromPath('graphite', array(
            'graphite_url' => urlencode($largeImgUrl),
            'graphite_iframe_w' => urlencode($this->iframeWidth),
            'graphite_iframe_h' => urlencode($this->iframeHeight)
        ));

        $html = '<a href="%s" title="%s"><img src="%s" alt="%s" width="300" height="120" /></a>';

        return sprintf(
            $html,
            $url,
            $metric,
            $imgUrl,
            $metric
      );
    }

    public function has(MonitoredObject $object)
    {
        if (($object instanceof Host)||($object instanceof Service)) {
            return true;
        } else {
            return false;
        }
    }

    public function getPreviewHtml(MonitoredObject $object)
    {
        $object->fetchCustomvars();

        if (array_key_exists("graphite", $object->customvars)) {
            $this->parseGrapherConfig($object->customvars["graphite"]);
        }

        $this->getKeysAndLabels(array(), $object->customvars);
        if (empty($this->graphiteKeys)) {
          $this->getPerfDataKeys($object);
        }

        if ($object instanceof Host) {
            $host = $object;
            $service = null;
        } elseif ($object instanceof Service) {
            $service = $object;
            $host = null;
        } else {
            return '';
        }

        $html = "<table class=\"avp newsection\">\n"
               ."<tbody>\n";

        for ($key = 0; $key < count($this->graphiteKeys); $key++) {
            $html .= "<tr><th>\n"
                  . $this->graphiteLabels[$key]
                  . '</th><td>'
                  . $this->getPreviewImage($host, $service, $this->graphiteKeys[$key])
                  . "</td>\n"
                  . "<tr>\n";
        }

        $html .= "</tbody></table>\n";
        return $html;
    }
}
